ipaserver/plugins/baseuser.py

```diff
@@ -728,3 +739,145 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
+        A list of values can be returned if --certificate is used multiple
+        times, or in conjunction with --subject --issuer.
+        """
+        data = []
+        data.extend(entry_attrs.get(cls.attribute, list()))
+
+        if 'issuer' in options or 'subject' in options:
+            issuer = options.get('issuer')
+            subject = options.get('subject')
+            data.append(cls._build_mapdata(subject, issuer))
+
+        for dercert in options.get('certificate', ()):
+            cert = x509.load_certificate(dercert, x509.DER)
+            issuer = DN(cert.issuer)
+            subject = DN(cert.subject)
+            data.append(cls._build_mapdata(subject, issuer))
+
+        entry_attrs[cls.attribute] = data
+
+    def get_args(self):
+        # ipacertmapdata is not mandatory as it can be built
+        # from the values subject+issuer or from reading certificate
+        for arg in super(ModCertMapData, self).get_args():
+            if arg.name == 'ipacertmapdata':
+                yield arg.clone(required=False, alwaysask=False)
+            else:
+                yield arg.clone()
+
+    def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys,
+                     **options):
+        # The 3 valid calls are
+        # ipa user-add-certmapdata LOGIN --subject xx --issuer yy
+        # ipa user-add-certmapdata LOGIN [DATA] --certificate xx
+        # ipa user-add-certmapdata LOGIN DATA
+        # Check that at least one of the 3 formats is used
+
+        try:
+            certmapdatas = keys[1] or []
+        except IndexError:
+            certmapdatas = []
+        issuer = options.get('issuer')
+        subject = options.get('subject')
+        certificates = options.get('certificate', [])
+
+        # If only LOGIN is supplied, then we need either subject or issuer or
+        # certificate
+        if not certmapdatas:
+            valid_options = (issuer, subject, certificates)
+            if all(not option for option in valid_options):
+                raise errors.RequirementError(name='CERTMAPDATA')
+
+        # If subject or issuer is provided, other options are not allowed
+        if subject or issuer:
+            if certificates:
+                raise errors.MutuallyExclusiveError(
+                    reason=_('cannot specify both subject/issuer '
+                             'and certificate'))
+            if certmapdatas:
+                raise errors.MutuallyExclusiveError(
+                    reason=_('cannot specify both subject/issuer '
+                             'and CERTMAPDATA'))
+
+        # if the command is called with --subject --issuer or --certificate
+        # we need to add ipacertmapdata to the attrs_list in order to
+        # display the resulting value in the command output
+        if 'ipacertmapdata' not in attrs_list:
+            attrs_list.append('ipacertmapdata')
+
+        self._convert_options_to_certmap(entry_attrs, options)
```

Comment:
Nitpick, but I think it would be better to pass `certmapdatas`, `issuer`, `subject` and `certificates` as arguments rather than `options`, so you can avoid the redundant `options.get()` etc. in `_convert_options_to_certmap()`.